src/codegen/ComponentConfig.ts

```diff
@@ -187,26 +263,215 @@ export class ComponentConfig extends GenerateComponentLike {
+        itemLine.push(`keyof ${base}`);
+        evalLines.push(`...props.${evaluator}(),`);
       }
+    }
+
+    const pluginInstances = this.plugins.map((plugin) => {
+      const args = plugin.makeConstructorArgs();
+      const instance = `new ${plugin.import}(${args})`;
+      return `'${plugin.getKey()}': ${instance}`;
+    });
+    const pluginMap = pluginInstances.length ? `protected plugins = {${pluginInstances.join(',\n')}};` : '';
+
+    function pluginRef(plugin: NodeDefPlugin<any>): string {
+      return `this.plugins['${plugin.getKey()}']`;
+    }
+
+    const pluginStateFactories = this.plugins
+      .filter((plugin) => plugin.stateFactory !== NodeDefPlugin.prototype.stateFactory)
+      .map((plugin) => `...${pluginRef(plugin)}.stateFactory(props as any),`)
+      .join('\n');
 
-      throw new Error(`Unknown capability ${key}`);
+    const pluginItemFactories = this.plugins
+      .filter((plugin) => plugin.itemFactory !== NodeDefPlugin.prototype.itemFactory)
+      .map((plugin) => `...${pluginRef(plugin)}.itemFactory(props as any)`)
+      .join(',\n');
+
+    const itemDef = pluginItemFactories
+      ? `const item = { ${pluginItemFactories} } as ${CompInternal}<'${this.type}'>;`
+      : '';
+
+    const pluginGeneratorChildren = this.plugins
+      .filter((plugin) => plugin.extraNodeGeneratorChildren !== NodeDefPlugin.prototype.extraNodeGeneratorChildren)
+      .map((plugin) => plugin.extraNodeGeneratorChildren())
+      .join('\n');
+
+    const additionalMethods: string[] = [];
+
+    if (!this.config.functionality.customExpressions) {
+      additionalMethods.push(
+        `// Do not override this one, set functionality.customExpressions to true instead
+        evalExpressions(props: ${ExprResolver}<'${this.type}'>) {
+          return this.evalDefaultExpressions(props);
+        }`,
+      );
+    }
+
+    if (this.hasDataModelBindings()) {
+      const LayoutValidationCtx = new CG.import({
+        import: 'LayoutValidationCtx',
+        from: 'src/features/devtools/layoutValidation/types',
+      });
+      additionalMethods.push(
+        `// You must implement this because the component has data model bindings defined
+        abstract validateDataModelBindings(ctx: ${LayoutValidationCtx}<'${this.type}'>): string[];`,
+      );
+    } else if (this.isFormLike()) {
+      additionalMethods.push(
+        `// This component could have, but does not have any data model bindings defined
+        getDisplayData() { return ''; }`,
+      );
+    }
+
+    for (const plugin of this.plugins) {
+      const extraMethodsFromPlugin = plugin.extraMethodsInDef();
+      additionalMethods.push(...extraMethodsFromPlugin);
+
+      const extraInEval = plugin.extraInEvalExpressions();
+      extraInEval && evalLines.push(extraInEval);
+    }
+
+    const childrenPlugins = this.plugins.filter((plugin) =>
+      isNodeDefChildrenPlugin(plugin),
+    ) as unknown as (NodeDefChildrenPlugin<any> & NodeDefPlugin<any>)[];
```

Comment:
I am curious if this could now work without a type assertion since we recently upgraded to TS 5.5, which has type inference on array.filter? 🤔

Reply:
Yup, it sure did! :star_struck: